crates/bevy_ecs/hecs/src/world.rs

```diff
@@ -904,6 +909,86 @@ impl From<MissingComponent> for ComponentError {
     }
 }
 
+#[cfg(feature = "dynamic-api")]
+use std::{cmp::Ordering, hash::Hasher};
+
+/// Uniquely identifies a type of component. This is conceptually similar to
+/// Rust's [`TypeId`], but allows for external type IDs to be defined.
+#[cfg(feature = "dynamic-api")]
+#[derive(Eq, PartialEq, Debug, Clone, Copy)]
+pub enum ComponentId {
```

Comment:
Let's newtype this with the internal at most `pub(crate)` to ensure the enumness doesn't leak out.

Reply:
Hmm, currently we use `ComponentId` to construct `DynamicQuery`s which ask for either Rust components or external components. Maybe we create a public enum that is used when making queries and then converted to the internal type when actually processing the query. The public type and the way you construct queries would then stay the same even if we change the internals.

Reply:
Is there a reason the public type needs to be an enum? An opaque struct that has a conversion and/or constructor defined iff dynamic components are enabled would have a lower surface area.

Reply:
Or, alternatively, distinct methods on `DynamicQuery` could be used.

Reply:
An upside to distinct methods is that the statically-typed version can encapsulate the existence of `TypeId`s entirely by taking a generic type parameter instead.

Reply:
OK, that sounds reasonable. I'll try to move all handling of component id out from the public interface and into respective functions on `DynamicQuery` and `EntityBuilder`.